indexing-service/src/main/java/org/apache/druid/indexing/overlord/supervisor/SupervisorManager.java

```diff
@@ -197,8 +197,17 @@ public boolean shouldUpdateSupervisor(SupervisorSpec spec)
       try {
         byte[] specAsBytes = jsonMapper.writeValueAsBytes(spec);
         Pair<Supervisor, SupervisorSpec> currentSupervisor = supervisors.get(spec.getId());
-        return currentSupervisor == null
-               || !Arrays.equals(specAsBytes, jsonMapper.writeValueAsBytes(currentSupervisor.rhs));
+        if (currentSupervisor == null || currentSupervisor.rhs == null) {
+          return true;
+        } else {
+          if (!Arrays.equals(specAsBytes, jsonMapper.writeValueAsBytes(currentSupervisor.rhs))) {
+            // The spec bytes are different, so we need to check if the replacement is allowed
+            currentSupervisor.rhs.validateSpecUpdateTo(spec);
+            return true;
+          } else {
+            return false;
+          }
+        }
```

Comment:
Cleaner with the nesting reduced and positive condition first:

```suggestion
        } else if (Arrays.equals(specAsBytes, jsonMapper.writeValueAsBytes(currentSupervisor.rhs))) {
          return false;
        } else {
          // The spec bytes are different, so we need to check if the update is allowed
          currentSupervisor.rhs.validateSpecUpdateTo(spec);
          return true;
        }
```